contributing/building-a-bento-video-player.md

```diff
@@ -0,0 +1,627 @@
+
+> ⚠️ Components may **not** embed scripts from a third-party location into host documents. If a third-party script is absolutely required, like on `<amp-ima-video>`, it must be inserted in an intermediate iframe, which we call a **proxy frame**.
+>
+> Proxy frames on Bento have not yet been tested as video player components, so they're not covered in this guide. If you wish to use one, please get in touch with `@alanorozco` via a Github issue or on Slack.
+
+### Forwarding `ref`
+
+To enable AMP actions (`my-element.play`) and the Preact component's imperative handle (`myPlayerRef.current.play()`), you'll have to [`forwardRef`](https://reactjs.org/docs/forwarding-refs.html). Rename `FantasticPlayer` to `FantasticPlayerWithRef`, and export a `FantasticPlayer` that forwards a `ref` into the former.
+
+```diff
++ import {forwardRef} from '../../../src/preact/compat';
+
+- export function FantasticPlayer({...rest}) {
++ function FantasticPlayerWithRef({...rest}, ref) {
+    ...
+  }
+
++ const FantasticPlayer = forwardRef(FantasticPlayerWithRef);
++ FantasticPlayer.displayName = 'FantasticPlayer'; // Make findable for tests.
++ export {FantasticPlayer};
+```
+
+So the outer structure looks like:
+
+```js
+// ...
+import {forwardRef} from '../../../src/preact/compat';
+
+// ...
+function FantasticPlayerWithRef({...rest}, ref) {
+  // ...
+}
+
+//...
+const FantasticPlayer = forwardRef(FantasticPlayerWithRef);
+FantasticPlayer.displayName = 'FantasticPlayer'; // Make findable for tests.
+export {FantasticPlayer};
+```
+
+### Loading an iframe with `VideoIframe`
+
+Your `FantasticPlayer` component should return a `VideoIframe` that's configured to a corresponding `postMessage` API. To start, we update the implementation in **`component.js`**:
+
+```diff
+- import {ContainWrapper} from '../../../src/preact/component';
++ import {VideoIframe} from '../../amp-video/1.0/video-iframe';
+
+  function FantasticPlayerWithRef({...rest}, ref) {
+-   ...
++   const src = useMemo(
++     () => 'https://example.com/fantastic',
++     []
++   );
++   const makeMethodMessage = useCallback(() => '{}', []);
++   const onMessage = useCallback((e) => {
++     console.log(e);
++   }, []);
+    return (
+-     <ContainWrapper layout size paint {...rest} >
+-       ...
+-     </ContainWrapper>
++     <VideoIframe
++       {...rest}
++       src={src}
++       makeMethodMessage={makeMethodMessage}
++       onMessage={onMessage}
++     />
+    );
+  }
+```
+
+into:
+
+```js
+import {VideoIframe} from '../../amp-video/1.0/video-iframe';
+
+// ...
+
+function FantasticPlayerWithRef({...rest}, ref) {
+  const src = useMemo(
+    () => 'https://example.com/fantastic',
+    []
+  );
+  const makeMethodMessage = useCallback(() => '{}', []);
+  const onMessage = useCallback((e) => {
+    console.log(e);
+  }, []);
+  return (
+    <VideoIframe
+      {...rest}
+      src={src}
+      makeMethodMessage={makeMethodMessage}
+      onMessage={onMessage}
```

Comment:
Syntax question: Would JSX destructuring allow you to just write `<VideoIframe {src, makeMethodMessage, onMesage, ...rest} />` here?

Reply:
`<Component {foo} />` is not part of the syntax, unfortunately.